### ISSUE TITLE
have esc after insert keep the cursor in place, no matter what?

### ISSUE THREAD

**Issue by ewtoombs** [mawww/kakoune#1309]
Except in replace / overwrite mode, I'd rather conceptualise the cursor as always being in between two characters. As such, it has always made more sense to me for the cursor to keep its position after an insert. When the cursor is in the middle of a line, this is exactly how it behaves, which is awesome, but when it's at the end of a line, it's still doing that little jump after esc is pressed. Is there a way to change this behaviour?

**Comment by mawww** [comment]
In Kakoune, conceptualising the cursor as being between two characters is an error, the cursors is always onto a character, as is the anchor, and they can eventually be on the same character, leading to a single character selection. If any of these could be between two characters, we could have empty selections, which is never the case in Kakoune.

That said, there is two things that can move the cursor on `<esc>`:
* Restoring cursor when exiting from an append insert, as if you hit `a` the cursor will move to next char for the insert session, we move it back one char at the end.
* Avoiding end of line. If the cursor is on an end of line when we exit insert mode, we go to previous character as well. This is mostly historical and there has been a movement towards avoiding end of lines in less and less cases, so this could be discussed further.

**Comment by mawww** [comment]
Hello, the selection cannot be empty because the whole design is based around cursor and anchors being *on* the character. For example, with an empty selection, the `d` key now either deletes nothing, or needs a special case to say 'if the selection is empty, delete the previous char'.

When I started working on Kakoune, I tried various ways to represent the selection, and the `anchor`, `cursor`, `on` characters is what worked best and sticked.

For eol avoidance, it'd be nice to have chat with the community about an eventual change there, I guess the irc channel (#kakoune on freenode) is the best place to float the idea and see what comes out.

**Comment by maximbaz** [comment]
Since the design of cursors is not going to change, should this issue be closed?

**Comment by ewtoombs** [comment]
Well, it's the only reason I don't use kakoune. I don't consider it enough of an improvement over vim without fixing this massive off-by-one error.

**Comment by maximbaz** [comment]
If I understand the issue correctly, this will be mitigated once we have different cursor shapes (#2194), i.e. if a cursor changed to a vertical bar, that would be exactly the behavior I see in nvim - correct?

![ins-cursor](https://user-images.githubusercontent.com/1177900/42472907-38408e5e-83c3-11e8-995b-f81eb4de98c3.gif)

**Comment by ewtoombs** [comment]
No, because <esc> after insert has inconsistent behaviour in kak. It behaves right when the cursor is in the middle of the line, but wrong when it is at the end.

**Comment by maximbaz** [comment]
Okay, then I guess I didn't understand the issue correctly, I don't know where to look but insert in the middle and in the end of the line seems pretty identical to me. Could you show a specific example, a gif maybe, or what exactly to press and what to expect?

**Comment by ewtoombs** [comment]
So, I am using kak again for the first time since I posted this. Looks like you guys have been busy! You fixed everything and didn't even notice! :D

It really did used to do what I had described in the first post, but it doesn't do that any more. Now, when I finish an `i` at the end of line, the cursor doesn't jump any more. That fixes the `xstextd` example I had given earlier.

So, looks like it's fixed! Closing now, and using kak for real this time. Молодец!
